hamocc/mo_parambgc_ini.F90

```diff
@@ -0,0 +1,1241 @@
+!                     A. Moree, C. Heinze
+!
+! This file is part of BLOM/iHAMOCC.
+!
+! BLOM is free software: you can redistribute it and/or modify it under the
+! terms of the GNU Lesser General Public License as published by the Free 
+! Software Foundation, either version 3 of the License, or (at your option) 
+! any later version. 
+!
+! BLOM is distributed in the hope that it will be useful, but WITHOUT ANY 
+! WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS 
+! FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for
+! more details. 
+!
+! You should have received a copy of the GNU Lesser General Public License 
+! along with BLOM. If not, see https://www.gnu.org/licenses/.
+
+module mo_parambgc_ini
+!******************************************************************************
+!
+! BELEG_PARM - now mo_parambgc_ini - initialize bgc parameters.
+!
+!  Ernst Maier-Reimer,    *MPI-Met, HH*    10.04.01
+!
+!  Modified
+!  --------
+!  J.Schwinger,        *NORCE Climate, Bergen*    2020-05-19
+!   -split the original BELEG_BGC in two parts, BELEG_PARM and BELEG_VARS
+!  jmaerz 
+!   - rename beleg_parm to mo_parambgc_ini 
+!
+!  Purpose
+!  -------
+!  - set bgc parameter values.
+!     
+!
+!  Parameter list:
+!  ---------------
+!  *INTEGER*   *kpie*    - 1st dimension of model grid.
+!  *INTEGER*   *kpje*    - 2nd dimension of model grid.
+!
+!******************************************************************************
+
+  use mo_carbch,      only: atm,atm_co2,atm_n2,atm_o2,dmspar 
+  use mo_biomod,      only: atten_c,atten_f,atten_uv,atten_w,bkopal,bkphy,bkopal,bkzoo,bluefix,ctochl,dremn2o,dremopal,        &
+                          & drempoc,dremsul,dyphy,ecan,epsher,fesoly,fetune,gammap,gammaz,grami,grazra,perc_diron,phytomi,     &
+                          & pi_alpha,rcalc,rcar, rdn2o1,rdn2o2,rdnit0,rdnit1,rdnit2,relaxfe,remido,riron,rnit,rnoi,ro2ut,      &
+                          & ropal,spemor,tf0,tf1,tf2,tff,wcal,wdust,wopal,wpoc,zinges 
+  use mo_sedmnt,      only: claydens,o2ut,rno3
+  use mo_control_bgc, only: io_stdo_bgc,bgc_namelist
+  use mo_param1_bgc,  only: iatmco2,iatmnco2,iatmo2,iatmn2,iatmc13,iatmc14,iatmbromo
+  use mod_xc,         only: mnproc
+
+#ifdef AGG
+  use mo_biomod,      only: alar1,alar2,alar3,alow1,alow2,alow3,calmax,cellmass,cellsink,dustd1,dustd2,dustd3,dustsink,        &
+                          & fractdim,fse,fsh,nmldmin,plower,pupper,safe,sinkexp,stick,tmfac,tsfac,vsmall,zdis
+#elif defined(WLIN)
+  use mo_biomod,      only: wmin,wmax,wlin
+#endif
+#ifdef BROMO
+  use mo_biomod,      only: rbro
+  use mo_carbch,      only: atm_bromo,fbro1,fbro2
+#endif
+#ifdef cisonew
+  use mo_biomod,      only: bifr13,bifr14,c14fac,prei13,prei14,re1312,re14to
+  use mo_carbch,      only: atm_c13, atm_c14,c14_t_half,c14dec
+#endif
+#ifdef natDIC
+  use mo_carbch,      only: atm_co2_nat
+#endif
+
+  implicit none
+
```

Comment:
A thought from my side: Having the variables that qualify as "parameters" defined in the module header of this module would simplify things a lot. Those variables that are truly "parameters" could be defined as such and assigned a value (that isn't changed), e.g.

  `integer, parameter :: rcar =122`

Parameters that can be modified via namelist could be assigned a default value, e.g.

  `real, save  :: atm_co2 = 284.32`

Then the problem of having to recalculate things after reading of the namelist would be obsolete, and the module would be much simpler (plus it would be much more transparent to the user - the fact that all parameter definition are scattered between different modules now isn't very nice in my opinion.

Of course this is more invasive and much more work (a lot of use statements would need to be changed).

Reply:
Hi @JorgSchwinger , seeing the upcoming pull request in #264 (and potentially my nitrogen cycle), I currently would opt for a step-wise approach, so first introducing this less invasive step, and after pulling in Marianas changes doing a second step and following your suggestion, potentially even go the step with the `protect` statement, what @TomasTorsvik suggested. Let's discuss this in person.